parser/src/parser.rs

```diff
@@ -1,46 +1,38 @@
 use crate::class::class_declaration;
 use flora_syntax::*;
-use nom::types::CompleteStr as Input;
-use nom::*;
+use nom::{combinator::map, multi::many0, IResult};
 
-named!(
-    pub declarations<Input, Declarations>,
-    many0!(declaration)
-);
+pub fn declarations(input: &str) -> IResult<&str, Declarations> {
+    many0(declaration)(input)
```

Comment:
I wonder why the code above passes the tests but not

```rust
separated_list(multispace0, declaration)(input)
```

This should have been the same as the old

```rust
named!(
    declaration<Input, Declaration>,
    ws!(alt!(
        map!(class_declaration, Declaration::Class)
    ))
);
```

Reply:
Not entirely sure what is going on, but the following would pass:
```rust
pub fn declarations(input: &str) -> IResult<&str, Declarations> {
    many0(declaration)(input)
}

fn declaration(input: &str) -> IResult<&str, Declaration> {
    delimited(
        multispace0,
        map(class_declaration, Declaration::Class),
        multispace0,
    )(input)
}
```

However, I think we should consciously decide what the right parser structure is.
With the current parser structure, both the superparser (`declaration`) and the subparser (`class_declaration`) allow ws. Because the parsers are greedy, in practice with the old code any ws before the first class declaration is consumed by `declaration`s leading `ws`, and any ws after a class declaration by `class_declaration`s trailing `multispace0`.

It would seem cleaner to me to have either only the superparser or only the subparser consume whitespace. The latter is the current state of this PR. The former seems more closely aligned with what humans would consider "part of the class declaration" vs "void between declarations". Maybe there is literature on whitespace in parser combinators?